src/vs/workbench/api/node/extHostTreeExplorers.ts

```diff
@@ -0,0 +1,101 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import { TreeExplorerNodeProvider } from 'vscode';
+import { TPromise } from 'vs/base/common/winjs.base';
+import { Disposable } from 'vs/workbench/api/node/extHostTypes';
+import { IThreadService } from 'vs/workbench/services/thread/common/threadService';
+import { MainContext, ExtHostTreeExplorersShape, MainThreadTreeExplorersShape } from './extHost.protocol';
+import { InternalTreeExplorerNode } from 'vs/workbench/parts/explorers/common/treeExplorerViewModel';
+import { ExtHostCommands } from 'vs/workbench/api/node/extHostCommands';
+import { asWinJsPromise } from 'vs/base/common/async';
+import { Severity } from 'vs/platform/message/common/message';
+
+export class ExtHostTreeExplorers extends ExtHostTreeExplorersShape {
+	private _proxy: MainThreadTreeExplorersShape;
+
+	private _treeExplorerNodeProviders: { [providerId: string]: TreeExplorerNodeProvider<any> };
+	private _externalNodeMaps: { [providerId: string]: { [id: number]: any } };
+
+	constructor(
+		threadService: IThreadService,
+		private commands: ExtHostCommands
+	) {
+		super();
+
+		this._proxy = threadService.get(MainContext.MainThreadExplorers);
+
+		this._treeExplorerNodeProviders = Object.create(null);
+		this._externalNodeMaps = Object.create(null);
+	}
+
+	registerTreeExplorerNodeProvider(providerId: string, provider: TreeExplorerNodeProvider<any>): Disposable {
```

Comment:
Shouldn't this return `vscode.Disposable` from `import * as vscode from 'vscode';`?